File: BaekjoonOnlineJudge/DFS/안전영역 (2468)/Solution.py
import copy
import sys
sys.setrecursionlimit(10000)

def dfs(province, rain, x, y):
  if x <= -1 or x >= N or y <= -1 or y >= N:
    return False
  if province[x][y] <= rain:
    return False
  province[x][y] = 0
  dfs(province, rain, x+1, y)
  dfs(province, rain, x-1, y)
  dfs(province, rain, x, y+1)
  dfs(province, rain, x, y-1)
  return True


N = int(input())
province = []
max_rain = 0

for _ in range(N):
  user_input = list(map(int, sys.stdin.readline().split()))
  max_rain = max(max_rain, max(user_input))
  province.append(user_input)

result = [1] # 강수량이 0일때의 경우를 미리 넣어둠
for rain in range(1, max_rain+1):
  cnt = 0
  temp = copy.deepcopy(province)
  for i in range(N):
    for j in range(N):
      if dfs(temp, rain, i, j) == True:
        cnt += 1
  result.append(cnt)

print(max(result))

File: BaekjoonOnlineJudge/DFS/안전영역 (2468)/Solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int N;
vector<vector<int>> graph;
vector<vector<int>> temp;

bool dfs(int x, int y,int height) {
	if (x <= -1 || x >= N || y <= -1 || y >= N || temp[x][y] <=height)
		return false;
	
	if (temp[x][y] > height) {
		temp[x][y] = height;

		dfs(x - 1, y, height);
		dfs(x, y - 1, height);
		dfs(x + 1, y, height);
		dfs(x, y + 1, height);

		return true;
	}
}

int main() {
	int result = 0;

	cin >> N;

	graph.assign(N, vector<int>(N, 0));
	temp.assign(N, vector<int>(N, 0));

	for (int i = 0;i < N;i++) {
		for (int j = 0;j < N;j++) {
			int height;
			cin >> height;
			graph[i][j] = height;
		}
	}

	for (int k = 0;k <= 100;k++) {
		int count = 0;
		copy(graph.begin(),graph.end(),temp.begin());
		for (int i = 0;i < N;i++) {
			for (int j = 0;j < N;j++) {
				if (dfs(i, j, k))
					count++;
			}
		}
		if (count>result)
			result = count;
	}

	cout << result;
}